src/main/java/com/google/cloud/tools/appengine/configuration/AppEngineWebXmlProjectStageConfiguration.java

```diff
@@ -127,8 +127,20 @@ public String getRuntime() {
     return runtime;
   }
 
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * @deprecated Use
+   *     newBuilder().sourceDirectory(sourceDirectory).stagingDirectory(stagingDirectory) instead.
```

Comment:
I'm not convinced this really changes the error situation.

Couldn't someone accidentally do `newBuilder().sourceDirectory(target).stagingDirectory(source)`?

Reply:
yes, but they are much less likely to do 

newBuilder().sourceDirectory(stagingDirectory).stagingDirectory(sourceDirectory)?

and if someone does it's easy to notice in code review. This is about making it harder to make a mistake, not making it impossible.

Reply:
FWIW I agree with @elharo's proposal.  The point of builders was to avoid multi-argument constructors… but we're now using multi-argument builders.  The downside is that missing-values are now a runtime error rather than a compile-time error.

Reply:
Should we use intermediate interfaces here for required params. I still don't like the idea of going from compile time checks to runtime checks.

I feel like someone (maybe @briandealwis) suggested something like this in the past.

Uncertain about these names, but just to make the point
```
public RequiresSourceDirectory newBuilder()

interface RequiresSourceDirectory {
  RequiresStagingDirectory sourceDirectory()
}

interface RequiresStagingDirectory {
  Builder stagingDirectory()
}

public static final class Builder implements RequiresSourceDirectory, RequiresStagingDirectory {
...
}
```

so any call to would *HAVE* to look like:
```
new Builder().sourceDirectory(sourceDir).stagingDirectory(stagingDir)....build()
```

Reply:
I don't think I follow. How does this require that those methods be invoked?

Reply:
Because you're only exposing the interface when you return the builder, you return the subset of functionality exposed by the interface.

newBuilder returns RequiresSourcedirectory 
sourceDirectory returns RequiresStagingDirectory
stagingDirectory returns Builder (for all optional things)

Reply:
Unless this is [auto-generated](https://github.com/google/auto/blob/master/value/userguide/builders.md), it sounds painful.

Reply:
@loosebazooka's proposal is clever. I'm still mulling it over in my head. I don't think autovalue will generate code like that though.

Reply:
Autovalue/builder actually creates code in the pattern shown in this PR.

Reply:
I think for simple cases like this it might work, but still looks ugly. Can punt on the idea for now.